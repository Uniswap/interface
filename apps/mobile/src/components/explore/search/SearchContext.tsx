export interface SearchContext {
  category?: string
  query?: string
  position?: number
  suggestionCount?: number
  isHistory?: boolean // history item click
}
