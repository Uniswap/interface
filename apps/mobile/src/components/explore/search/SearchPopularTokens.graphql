query SearchPopularTokens {
  topTokens(chain: ETHEREUM, orderBy: VOLUME, page: 1, pageSize: 2) {
    id
    address
    chain
    symbol
    decimals
    project {
      id
      name
      logoUrl
      safetyLevel
    }
  }
  # `topTokens` returns WETH rather than ETH
  # here we retrieve ETH information to swap out in the UI
  eth: tokens(contracts: [{ address: null, chain: ETHEREUM }]) {
    id
    address
    chain
    symbol
    decimals
    project {
      id
      name
      logoUrl
      safetyLevel
    }
  }
}
