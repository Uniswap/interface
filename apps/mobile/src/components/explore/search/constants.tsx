export const SEARCH_RESULT_HEADER_KEY = 'header'
