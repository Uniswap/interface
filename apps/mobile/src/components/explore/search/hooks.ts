import { useMemo } from 'react'
import { useEnabledChains } from 'uniswap/src/features/chains/hooks/useEnabledChains'
import { UniverseChainId } from 'uniswap/src/features/chains/types'
import { ENS_SUFFIX } from 'uniswap/src/features/ens/constants'
import { useENS } from 'uniswap/src/features/ens/useENS'
import { SearchResultType, WalletSearchResult } from 'uniswap/src/features/search/SearchResult'
import { useUnitagByAddress, useUnitagByName } from 'uniswap/src/features/unitags/hooks'
import { getValidAddress } from 'uniswap/src/utils/addresses'
import { useIsSmartContractAddress } from 'wallet/src/features/transactions/send/hooks/useIsSmartContractAddress'

// eslint-disable-next-line complexity
export function useWalletSearchResults(
  query: string,
  selectedChain: UniverseChainId | null,
): {
  wallets: WalletSearchResult[]
  loading: boolean
  exactENSMatch: boolean
  exactUnitagMatch: boolean
} {
  const { defaultChainId } = useEnabledChains()
  const validAddress: Address | undefined = useMemo(() => getValidAddress(query, true, false) ?? undefined, [query])

  const querySkippedIfValidAddress = validAddress ? null : query

  // Search for matching .eth if not a valid address
  const {
    address: dotEthAddress,
    name: dotEthName,
    loading: dotEthLoading,
  } = useENS({ nameOrAddress: querySkippedIfValidAddress, autocompleteDomain: true })

  // Search for exact match for ENS if not a valid address
  const {
    address: ensAddress,
    name: ensName,
    loading: ensLoading,
  } = useENS({ nameOrAddress: querySkippedIfValidAddress, autocompleteDomain: false })

  // Search for matching Unitag by name
  const { unitag: unitagByName, loading: unitagLoading } = useUnitagByName(query)

  // Search for matching Unitag by address
  const { unitag: unitagByAddress, loading: unitagByAddressLoading } = useUnitagByAddress(validAddress)

  // Search for matching EOA wallet address
  const { isSmartContractAddress, loading: loadingIsSmartContractAddress } = useIsSmartContractAddress(
    validAddress,
    selectedChain ?? defaultChainId,
  )

  const hasENSResult = dotEthName && dotEthAddress
  const hasEOAResult = validAddress && !isSmartContractAddress

  // Consider when to show sections

  // Only consider queries with the .eth suffix as an exact ENS match
  const exactENSMatch = dotEthName?.toLowerCase() === query.toLowerCase() && query.includes(ENS_SUFFIX)

  const results: WalletSearchResult[] = []

  // Prioritize unitags

  if (unitagByName?.address?.address && unitagByName?.username) {
    results.push({
      type: SearchResultType.Unitag,
      address: unitagByName.address.address,
      unitag: unitagByName.username,
    })
  }

  // Add full address if relevant
  if (unitagByAddress?.username && validAddress) {
    results.push({
      type: SearchResultType.Unitag,
      address: validAddress,
      unitag: unitagByAddress.username,
    })
  }

  // Add the raw ENS result if available and a unitag by address was not already added
  if (!validAddress && ensAddress && ensName && !unitagByAddress?.username) {
    results.push({
      type: SearchResultType.ENSAddress,
      address: ensAddress,
      ensName,
      isRawName: !ensName.endsWith(ENS_SUFFIX), // Ensure raw name is used for subdomains only
    })
  }

  // Add ENS result if it's different than the unitag result and raw ENS result
  if (
    !validAddress &&
    hasENSResult &&
    dotEthAddress !== unitagByName?.address?.address &&
    dotEthAddress !== ensAddress
  ) {
    results.push({
      type: SearchResultType.ENSAddress,
      address: dotEthAddress,
      ensName: dotEthName,
    })
  }

  // Do not show EOA address result if there is a Unitag result by address
  if (hasEOAResult && !unitagByAddress?.username) {
    results.push({
      type: SearchResultType.WalletByAddress,
      address: validAddress,
    })
  }

  // Ensure loading is returned
  const walletsLoading =
    dotEthLoading || ensLoading || loadingIsSmartContractAddress || unitagLoading || unitagByAddressLoading

  return {
    loading: walletsLoading,
    wallets: results,
    exactENSMatch,
    exactUnitagMatch: !!(unitagByName || unitagByAddress),
  }
}
