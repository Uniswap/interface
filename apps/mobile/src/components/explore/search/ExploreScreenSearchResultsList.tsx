import { memo, useCallback, useState } from 'react'
import { KeyboardAvoidingView, TextInput } from 'react-native'
import { FlatList } from 'react-native-gesture-handler'
import { SearchEmptySection } from 'src/components/explore/search/SearchEmptySection'
import { SearchResultsSection } from 'src/components/explore/search/SearchResultsSection'
import { Flex, Text, TouchableArea, flexStyles } from 'ui/src'

import { UniverseChainId } from 'uniswap/src/features/chains/types'
import { FeatureFlags } from 'uniswap/src/features/gating/flags'
import { useFeatureFlag } from 'uniswap/src/features/gating/hooks'
import { SearchModalNoQueryList } from 'uniswap/src/features/search/SearchModal/SearchModalNoQueryList'
import { SearchModalResultsList } from 'uniswap/src/features/search/SearchModal/SearchModalResultsList'
import { MOBILE_SEARCH_TABS, SearchTab } from 'uniswap/src/features/search/SearchModal/types'
import Trace from 'uniswap/src/features/telemetry/Trace'
import { SectionName } from 'uniswap/src/features/telemetry/constants'
import { useDebounce } from 'utilities/src/time/timing'

function LegacyExploreSearchResultsList({
  debouncedSearchQuery,
  chainFilter,
  textInputRef,
}: {
  debouncedSearchQuery: string
  chainFilter: UniverseChainId | null
  textInputRef: React.RefObject<TextInput>
}): JSX.Element {
  const onScroll = useCallback(() => {
    textInputRef.current?.blur()
  }, [textInputRef])

  return (
    <>
      <Flex p="$spacing4" />
      {debouncedSearchQuery.length === 0 ? (
        // Mimic ScrollView behavior with FlatList
        // Needs to be from gesture handler to work on android within BottomSheelModal
        <FlatList
          ListHeaderComponent={<SearchEmptySection selectedChain={chainFilter} />}
          data={[]}
          keyExtractor={(): string => 'search-empty-section-container'}
          keyboardShouldPersistTaps="always"
          renderItem={null}
          scrollEventThrottle={16}
          showsHorizontalScrollIndicator={false}
          showsVerticalScrollIndicator={false}
          onScroll={onScroll}
        />
      ) : (
        <SearchResultsSection searchQuery={debouncedSearchQuery} selectedChain={chainFilter} />
      )}
    </>
  )
}

function NewExploreSearchResultsList({
  searchQuery,
  chainFilter,
  debouncedSearchQuery,
  debouncedParsedSearchQuery,
}: {
  searchQuery: string
  chainFilter: UniverseChainId | null
  debouncedSearchQuery: string | null
  debouncedParsedSearchQuery: string | null
}): JSX.Element {
  const [activeTab, setActiveTab] = useState<SearchTab>(SearchTab.All)

  return (
    <Trace section={SectionName.ExploreSearch}>
      <Flex row px="$spacing20" pt="$spacing16" pb="$spacing8" gap="$spacing16">
        {MOBILE_SEARCH_TABS.map((tab) => (
          <TouchableArea key={tab} onPress={() => setActiveTab(tab)}>
            <Text color={activeTab === tab ? '$neutral1' : '$neutral2'} variant="buttonLabel2">
              {tab}
            </Text>
          </TouchableArea>
        ))}
      </Flex>
      {searchQuery && searchQuery.length > 0 ? (
        <SearchModalResultsList
          chainFilter={chainFilter}
          debouncedParsedSearchFilter={debouncedParsedSearchQuery}
          debouncedSearchFilter={debouncedSearchQuery}
          searchFilter={searchQuery}
          activeTab={activeTab}
        />
      ) : (
        <SearchModalNoQueryList chainFilter={chainFilter} activeTab={activeTab} />
      )}
    </Trace>
  )
}

export const ExploreScreenSearchResultsList = memo(function _ExploreScreenSearchResultsList({
  searchQuery,
  parsedSearchQuery,
  chainFilter,
  textInputRef,
}: {
  searchQuery: string
  parsedSearchQuery: string | null
  chainFilter: UniverseChainId | null
  textInputRef: React.RefObject<TextInput>
}): JSX.Element {
  const searchRevampEnabled = useFeatureFlag(FeatureFlags.SearchRevamp)

  const debouncedSearchQuery = useDebounce(searchQuery)
  const debouncedParsedSearchQuery = useDebounce(parsedSearchQuery)

  return (
    <KeyboardAvoidingView behavior="height" style={flexStyles.fill}>
      {searchRevampEnabled ? (
        <NewExploreSearchResultsList
          searchQuery={searchQuery}
          chainFilter={chainFilter}
          debouncedSearchQuery={debouncedSearchQuery}
          debouncedParsedSearchQuery={debouncedParsedSearchQuery}
        />
      ) : (
        <LegacyExploreSearchResultsList
          debouncedSearchQuery={debouncedParsedSearchQuery ?? debouncedSearchQuery ?? ''}
          chainFilter={chainFilter}
          textInputRef={textInputRef}
        />
      )}
    </KeyboardAvoidingView>
  )
})
