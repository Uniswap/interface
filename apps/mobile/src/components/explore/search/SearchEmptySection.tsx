import React from 'react'
import { useTranslation } from 'react-i18next'
import { FlatList } from 'react-native'
import { FadeIn, FadeOut } from 'react-native-reanimated'
import { useAppDispatch, useAppSelector } from 'src/app/hooks'
import { SearchPopularNFTCollections } from 'src/components/explore/search/SearchPopularNFTCollections'
import { SearchPopularTokens } from 'src/components/explore/search/SearchPopularTokens'
import { renderSearchItem } from 'src/components/explore/search/SearchResultsSection'
import { SectionHeaderText } from 'src/components/explore/search/SearchSectionHeader'
import { clearSearchHistory } from 'src/features/explore/searchHistorySlice'
import { SearchResultType, WalletSearchResult } from 'src/features/explore/SearchResult'
import { selectSearchHistory } from 'src/features/explore/selectSearchHistory'
import { AnimatedFlex, Flex, Text, TouchableArea, useSporeColors } from 'ui/src'
import ClockIcon from 'ui/src/assets/icons/clock.svg'
import TrendArrowIcon from 'ui/src/assets/icons/trend-up.svg'
import { iconSizes } from 'ui/src/theme'

export const SUGGESTED_WALLETS: WalletSearchResult[] = [
  {
    type: SearchResultType.Wallet,
    address: '0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045',
    ensName: 'vitalik.eth',
  },
  {
    type: SearchResultType.Wallet,
    address: '0x50EC05ADe8280758E2077fcBC08D878D4aef79C3',
    ensName: 'hayden.eth',
  },
]

export function SearchEmptySection(): JSX.Element {
  const { t } = useTranslation()
  const dispatch = useAppDispatch()
  const searchHistory = useAppSelector(selectSearchHistory)

  const onPressClearSearchHistory = (): void => {
    dispatch(clearSearchHistory())
  }

  // Show search history (if applicable), trending tokens, and wallets
  return (
    <AnimatedFlex entering={FadeIn} exiting={FadeOut} gap="$spacing12">
      {searchHistory.length > 0 && (
        <AnimatedFlex entering={FadeIn} exiting={FadeOut}>
          <FlatList
            ListHeaderComponent={
              <Flex
                row
                alignItems="center"
                gap="$spacing16"
                justifyContent="space-between"
                mb="$spacing4">
                <SectionHeaderText icon={<RecentIcon />} title={t('Recent searches')} />
                <TouchableArea onPress={onPressClearSearchHistory}>
                  <Text color="$accent1" variant="buttonLabel3">
                    {t('Clear all')}
                  </Text>
                </TouchableArea>
              </Flex>
            }
            data={searchHistory}
            renderItem={(props): JSX.Element | null =>
              renderSearchItem({ ...props, searchContext: { isHistory: true } })
            }
          />
        </AnimatedFlex>
      )}
      <Flex gap="$spacing4">
        <SectionHeaderText icon={<TrendIcon />} title={t('Popular tokens')} />
        <SearchPopularTokens />
      </Flex>
      <Flex gap="$spacing4">
        <SectionHeaderText icon={<TrendIcon />} title={t('Popular NFT collections')} />
        <SearchPopularNFTCollections />
      </Flex>
      <FlatList
        ListHeaderComponent={
          <SectionHeaderText icon={<TrendIcon />} title={t('Suggested wallets')} />
        }
        data={SUGGESTED_WALLETS}
        keyExtractor={walletKey}
        renderItem={renderSearchItem}
      />
    </AnimatedFlex>
  )
}

const walletKey = (wallet: WalletSearchResult): string => {
  return wallet.address
}

export const TrendIcon = (): JSX.Element => {
  const colors = useSporeColors()
  return <TrendArrowIcon color={colors.neutral2.get()} width={iconSizes.icon20} />
}

export const RecentIcon = (): JSX.Element => {
  const colors = useSporeColors()
  return (
    <ClockIcon color={colors.neutral2.get()} height={iconSizes.icon20} width={iconSizes.icon20} />
  )
}
