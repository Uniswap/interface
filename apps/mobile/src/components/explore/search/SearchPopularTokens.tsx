import React, { useMemo } from 'react'
import { FlatList, ListRenderItemInfo } from 'react-native'
import { SearchTokenItem } from 'src/components/explore/search/items/SearchTokenItem'
import { getSearchResultId } from 'src/components/explore/search/utils'
import { Loader } from 'src/components/loading'
import { SearchResultType, TokenSearchResult } from 'src/features/explore/SearchResult'
import { TopToken, usePopularTokens } from 'src/features/tokens/hooks'
import { Inset } from 'ui/src'
import { fromGraphQLChain } from 'wallet/src/features/chains/utils'

function gqlTokenToTokenSearchResult(token: Maybe<TopToken>): TokenSearchResult | null {
  if (!token || !token.project) return null

  const { chain, address, symbol, project } = token
  const { name } = project
  const chainId = fromGraphQLChain(chain)
  if (!chainId || !symbol || !name) return null

  return {
    type: SearchResultType.Token,
    chainId,
    address,
    name,
    symbol,
    logoUrl: project?.logoUrl,
  } as TokenSearchResult
}

export function SearchPopularTokens(): JSX.Element {
  const { popularTokens, loading } = usePopularTokens()
  const tokens = useMemo(
    () =>
      popularTokens
        ?.map(gqlTokenToTokenSearchResult)
        .filter((t): t is TokenSearchResult => Boolean(t)),
    [popularTokens]
  )

  if (loading) {
    return (
      <Inset all="$spacing8">
        <Loader.Token repeat={2} />
      </Inset>
    )
  }

  return <FlatList data={tokens} keyExtractor={getSearchResultId} renderItem={renderTokenItem} />
}

const renderTokenItem = ({ item }: ListRenderItemInfo<TokenSearchResult>): JSX.Element => (
  <SearchTokenItem token={item} />
)
