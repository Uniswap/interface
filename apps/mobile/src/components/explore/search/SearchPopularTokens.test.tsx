import React from 'react'
import { SearchPopularTokens } from 'src/components/explore/search/SearchPopularTokens'
import { EthToken, TopNFTCollections, TopTokens } from 'src/test/gqlFixtures'
import { render, screen } from 'src/test/test-utils'
import {
  SearchPopularNftCollectionsDocument,
  SearchPopularTokensDocument,
} from 'wallet/src/data/__generated__/types-and-hooks'

const TokensMock = {
  request: {
    query: SearchPopularTokensDocument,
  },
  result: {
    data: {
      topTokens: TopTokens,
      eth: EthToken,
    },
  },
}

const NFTsMock = {
  request: {
    query: SearchPopularNftCollectionsDocument,
  },
  result: {
    data: {
      topCollections: TopNFTCollections,
    },
  },
}

describe(SearchPopularTokens, () => {
  it('renders without error', async () => {
    const tree = render(<SearchPopularTokens />, { mocks: [TokensMock, NFTsMock] })

    // Loading should show Token loader
    expect(screen.getAllByText('Token Full Name')).toBeDefined()
    expect(tree.toJSON()).toMatchSnapshot()

    // Success where WETH result in topTokens is replaced by ETH
    expect(await screen.findByText('ETH')).toBeDefined()
    expect(screen.getByText('USDC')).toBeDefined()
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
