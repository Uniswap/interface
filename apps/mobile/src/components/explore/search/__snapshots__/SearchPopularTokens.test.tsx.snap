// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPopularTokens renders without error 1`] = `
<View
  style={
    {
      "alignItems": "stretch",
      "flexDirection": "column",
      "paddingBottom": 8,
      "paddingLeft": 8,
      "paddingRight": 8,
      "paddingTop": 8,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      {
        "alignItems": "stretch",
        "flexDirection": "column",
        "opacity": 0,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "stretch",
          "flexDirection": "column",
          "flexGrow": 1,
          "gap": 4,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "minHeight": 56,
            "opacity": 1,
            "paddingBottom": 8,
            "paddingTop": 8,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "flexShrink": 1,
              "gap": 12,
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              {
                "alignItems": "stretch",
                "backgroundColor": "#CECECE",
                "borderBottomLeftRadius": 999999,
                "borderBottomRightRadius": 999999,
                "borderTopLeftRadius": 999999,
                "borderTopRightRadius": 999999,
                "flexDirection": "column",
                "minHeight": 40,
                "minWidth": 40,
              }
            }
          />
          <View
            style={
              {
                "alignItems": "flex-start",
                "flexDirection": "column",
                "flexShrink": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "position": "relative",
                  }
                }
              >
                <View
                  accessibilityElementsHidden={true}
                  importantForAccessibility="no-hide-descendants"
                >
                  <Text
                    allowFontScaling={true}
                    maxFontSizeMultiplier={1.4}
                    numberOfLines={1}
                    style={
                      {
                        "color": "transparent",
                        "fontFamily": "Basel-Book",
                        "fontSize": 19,
                        "lineHeight": 24,
                        "opacity": 0,
                      }
                    }
                    suppressHighlighting={true}
                  >
                    Token Full Name
                  </Text>
                </View>
                <View
                  style={
                    {
                      "alignItems": "stretch",
                      "backgroundColor": "#F9F9F9",
                      "borderBottomLeftRadius": 4,
                      "borderBottomRightRadius": 4,
                      "borderTopLeftRadius": 4,
                      "borderTopRightRadius": 4,
                      "bottom": "5%",
                      "flexDirection": "column",
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": "5%",
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 8,
                  "minHeight": 20,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "position": "relative",
                    }
                  }
                >
                  <View
                    accessibilityElementsHidden={true}
                    importantForAccessibility="no-hide-descendants"
                  >
                    <Text
                      allowFontScaling={true}
                      maxFontSizeMultiplier={1.4}
                      numberOfLines={1}
                      style={
                        {
                          "color": "transparent",
                          "fontFamily": "Basel-Book",
                          "fontSize": 17,
                          "lineHeight": 24,
                          "opacity": 0,
                        }
                      }
                      suppressHighlighting={true}
                    >
                      1,000 TFN
                    </Text>
                  </View>
                  <View
                    style={
                      {
                        "alignItems": "stretch",
                        "backgroundColor": "#F9F9F9",
                        "borderBottomLeftRadius": 4,
                        "borderBottomRightRadius": 4,
                        "borderTopLeftRadius": 4,
                        "borderTopRightRadius": 4,
                        "bottom": "5%",
                        "flexDirection": "column",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": "5%",
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "minHeight": 56,
            "opacity": 0.5,
            "paddingBottom": 8,
            "paddingTop": 8,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "flexShrink": 1,
              "gap": 12,
              "overflow": "hidden",
            }
          }
        >
          <View
            style={
              {
                "alignItems": "stretch",
                "backgroundColor": "#CECECE",
                "borderBottomLeftRadius": 999999,
                "borderBottomRightRadius": 999999,
                "borderTopLeftRadius": 999999,
                "borderTopRightRadius": 999999,
                "flexDirection": "column",
                "minHeight": 40,
                "minWidth": 40,
              }
            }
          />
          <View
            style={
              {
                "alignItems": "flex-start",
                "flexDirection": "column",
                "flexShrink": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "position": "relative",
                  }
                }
              >
                <View
                  accessibilityElementsHidden={true}
                  importantForAccessibility="no-hide-descendants"
                >
                  <Text
                    allowFontScaling={true}
                    maxFontSizeMultiplier={1.4}
                    numberOfLines={1}
                    style={
                      {
                        "color": "transparent",
                        "fontFamily": "Basel-Book",
                        "fontSize": 19,
                        "lineHeight": 24,
                        "opacity": 0,
                      }
                    }
                    suppressHighlighting={true}
                  >
                    Token Full Name
                  </Text>
                </View>
                <View
                  style={
                    {
                      "alignItems": "stretch",
                      "backgroundColor": "#F9F9F9",
                      "borderBottomLeftRadius": 4,
                      "borderBottomRightRadius": 4,
                      "borderTopLeftRadius": 4,
                      "borderTopRightRadius": 4,
                      "bottom": "5%",
                      "flexDirection": "column",
                      "left": 0,
                      "position": "absolute",
                      "right": 0,
                      "top": "5%",
                    }
                  }
                />
              </View>
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "gap": 8,
                  "minHeight": 20,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "position": "relative",
                    }
                  }
                >
                  <View
                    accessibilityElementsHidden={true}
                    importantForAccessibility="no-hide-descendants"
                  >
                    <Text
                      allowFontScaling={true}
                      maxFontSizeMultiplier={1.4}
                      numberOfLines={1}
                      style={
                        {
                          "color": "transparent",
                          "fontFamily": "Basel-Book",
                          "fontSize": 17,
                          "lineHeight": 24,
                          "opacity": 0,
                        }
                      }
                      suppressHighlighting={true}
                    >
                      1,000 TFN
                    </Text>
                  </View>
                  <View
                    style={
                      {
                        "alignItems": "stretch",
                        "backgroundColor": "#F9F9F9",
                        "borderBottomLeftRadius": 4,
                        "borderBottomRightRadius": 4,
                        "borderTopLeftRadius": 4,
                        "borderTopRightRadius": 4,
                        "bottom": "5%",
                        "flexDirection": "column",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": "5%",
                      }
                    }
                  />
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`SearchPopularTokens renders without error 2`] = `
<RCTScrollView
  data={
    [
      {
        "address": null,
        "chainId": 1,
        "logoUrl": "https://loremflickr.com/640/480",
        "name": "Ethereum",
        "symbol": "ETH",
        "type": 1,
      },
      {
        "address": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
        "chainId": 1,
        "logoUrl": "https://loremflickr.com/640/480",
        "name": "USD Coin",
        "symbol": "USDC",
        "type": 1,
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        actions={
          [
            {
              "onPress": [Function],
              "systemIcon": "heart.fill",
              "title": "Remove favorite",
            },
            {
              "onPress": [Function],
              "systemIcon": "arrow.2.squarepath",
              "title": "Swap",
            },
            {
              "onPress": [Function],
              "systemIcon": "square.and.arrow.up",
              "title": "Share",
            },
          ]
        }
        onPress={[Function]}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            {
              "bottom": 5,
              "left": 5,
              "right": 5,
              "top": 5,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="search-token-item"
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "gap": 12,
                "paddingBottom": 12,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 12,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "column",
                  "height": 40,
                  "justifyContent": "center",
                  "width": 40,
                }
              }
            >
              <Image
                source={
                  {
                    "uri": "https://loremflickr.com/640/480",
                  }
                }
                style={
                  [
                    {
                      "resizeMode": "contain",
                    },
                    {
                      "backgroundColor": "#2222220D",
                      "borderColor": "#2222220D",
                      "borderRadius": 20,
                      "borderWidth": 0.5,
                      "height": 40,
                      "width": 40,
                    },
                  ]
                }
              />
            </View>
            <View
              style={
                {
                  "alignItems": "flex-start",
                  "flexDirection": "column",
                  "flexShrink": 1,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 8,
                    "justifyContent": "center",
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "stretch",
                      "flexDirection": "column",
                      "flexShrink": 1,
                    }
                  }
                >
                  <Text
                    allowFontScaling={true}
                    maxFontSizeMultiplier={1.4}
                    numberOfLines={1}
                    style={
                      {
                        "color": "#222222",
                        "fontFamily": "Basel-Book",
                        "fontSize": 19,
                        "lineHeight": 24,
                      }
                    }
                    suppressHighlighting={true}
                  >
                    Ethereum
                  </Text>
                </View>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 8,
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  allowFontScaling={true}
                  maxFontSizeMultiplier={1.4}
                  numberOfLines={1}
                  style={
                    {
                      "color": "#7D7D7D",
                      "fontFamily": "Basel-Book",
                      "fontSize": 17,
                      "lineHeight": 24,
                    }
                  }
                  suppressHighlighting={true}
                >
                  ETH
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        actions={
          [
            {
              "onPress": [Function],
              "systemIcon": "heart",
              "title": "Favorite token",
            },
            {
              "onPress": [Function],
              "systemIcon": "arrow.2.squarepath",
              "title": "Swap",
            },
            {
              "onPress": [Function],
              "systemIcon": "square.and.arrow.up",
              "title": "Share",
            },
          ]
        }
        onPress={[Function]}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          hitSlop={
            {
              "bottom": 5,
              "left": 5,
              "right": 5,
              "top": 5,
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="search-token-item"
        >
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "gap": 12,
                "paddingBottom": 12,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 12,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "column",
                  "height": 40,
                  "justifyContent": "center",
                  "width": 40,
                }
              }
            >
              <Image
                source={
                  {
                    "uri": "https://loremflickr.com/640/480",
                  }
                }
                style={
                  [
                    {
                      "resizeMode": "contain",
                    },
                    {
                      "backgroundColor": "#2222220D",
                      "borderColor": "#2222220D",
                      "borderRadius": 20,
                      "borderWidth": 0.5,
                      "height": 40,
                      "width": 40,
                    },
                  ]
                }
              />
            </View>
            <View
              style={
                {
                  "alignItems": "flex-start",
                  "flexDirection": "column",
                  "flexShrink": 1,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 8,
                    "justifyContent": "center",
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "stretch",
                      "flexDirection": "column",
                      "flexShrink": 1,
                    }
                  }
                >
                  <Text
                    allowFontScaling={true}
                    maxFontSizeMultiplier={1.4}
                    numberOfLines={1}
                    style={
                      {
                        "color": "#222222",
                        "fontFamily": "Basel-Book",
                        "fontSize": 19,
                        "lineHeight": 24,
                      }
                    }
                    suppressHighlighting={true}
                  >
                    USD Coin
                  </Text>
                </View>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 8,
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  allowFontScaling={true}
                  maxFontSizeMultiplier={1.4}
                  numberOfLines={1}
                  style={
                    {
                      "color": "#7D7D7D",
                      "fontFamily": "Basel-Book",
                      "fontSize": 17,
                      "lineHeight": 24,
                    }
                  }
                  suppressHighlighting={true}
                >
                  USDC
                </Text>
                <View
                  style={
                    {
                      "alignItems": "stretch",
                      "flexDirection": "column",
                      "flexShrink": 1,
                    }
                  }
                >
                  <Text
                    allowFontScaling={true}
                    maxFontSizeMultiplier={1.4}
                    numberOfLines={1}
                    style={
                      {
                        "color": "#7D7D7D",
                        "fontFamily": "Basel-Book",
                        "fontSize": 17,
                        "lineHeight": 24,
                      }
                    }
                    suppressHighlighting={true}
                  >
                    0xA0b8...eB48
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
